prov/verbs_ud/src/verbs_ud_ep.c

```diff
@@ -0,0 +1,807 @@
+	int ret = FI_SUCCESS;
+	struct ofi_verbs_ud_wr_entry *wr_entry;
+
+	fastlock_acquire(&util_cq->cq_lock);
+	if (unlikely(ofi_cirque_isfull(util_cq->cirq))) {
+		VERBS_UD_DBG(FI_LOG_CQ, "util_cq cirq is full!\n");
+		ret = -FI_EAGAIN;
+		goto out;
+	}
+
+	wr_entry = (struct ofi_verbs_ud_wr_entry *)(uintptr_t)wc->wr_id;
+
+	comp = ofi_cirque_tail(util_cq->cirq);
+	comp->op_context = wr_entry->hdr.context;
+	comp->flags = wr_entry->hdr.flags;
+	comp->len = (uint64_t) wc->byte_len;
+	comp->buf = NULL;
+	comp->data = (wc->wc_flags & IBV_WC_WITH_IMM) ?
+		ntohl(wc->imm_data) : 0;
+	comp->tag = 0;
+	ofi_cirque_commit(util_cq->cirq);
+	ofi_verbs_ud_wr_entry_release(
+		&wr_entry->hdr.ep->grh_pool,
+		(struct ofi_verbs_ud_wr_entry_hdr *)wr_entry
+	);
+out:
+	fastlock_release(&util_cq->cq_lock);
+	return ret;
+}
+
+static inline
+void ofi_verbs_ud_cq_report_error(struct util_cq *util_cq,
+				  struct ibv_wc *wc)
+{
+	struct fi_cq_err_entry err_entry = {
+		.err = EIO,
+		.prov_errno = wc->status,
+		.err_data_size = sizeof(wc->vendor_err),
+		.err_data = (void *)&wc->vendor_err,
+	};
+	struct fi_cq_tagged_entry *comp;
+	struct ofi_verbs_ud_wr_entry *wr_entry =
+		(struct ofi_verbs_ud_wr_entry *)(uintptr_t)wc->wr_id;
+	struct util_cq_err_entry *err = calloc(1, sizeof(*err));
+	if (!err) {
+		VERBS_UD_WARN(FI_LOG_CQ,
+			      "Unable to allocate "
+			      "util_cq_err_entry\n");
+		goto out;
+	}
+
+	err_entry.op_context = wr_entry->hdr.context;
+	err_entry.flags = wr_entry->hdr.flags;
+	err->err_entry = err_entry;
+
+	fastlock_acquire(&util_cq->cq_lock);
+	slist_insert_tail(&err->list_entry, &util_cq->err_list);
+
+	/* Signal that there is err entry */
+	comp = ofi_cirque_tail(util_cq->cirq);
+	comp->flags = UTIL_FLAG_ERROR;
+	ofi_cirque_commit(util_cq->cirq);
+
+	fastlock_release(&util_cq->cq_lock);
+
+out:
+	ofi_verbs_ud_wr_entry_release(
+		&wr_entry->hdr.ep->grh_pool,
+		(struct ofi_verbs_ud_wr_entry_hdr *)wr_entry
+	);
+}
+
+static inline
+void ofi_verbs_ud_ep_recv_progress(struct util_ep *util_ep)
+{
+	struct ofi_verbs_ud_ep *ep;
+	struct ibv_wc wc = (const struct ibv_wc) { 0 };
+	int ret;
+
+	ep = container_of(util_ep, struct ofi_verbs_ud_ep, util_ep);
+
+	ret = ibv_poll_cq(ep->ibv_recv_cq, 1, &wc);
+	if (ret > 0) {
+		if (likely(wc.status == IBV_WC_SUCCESS))
+			(void) ofi_verbs_ud_cq_comp(util_ep->rx_cq, &wc);
+		else
+			ofi_verbs_ud_cq_report_error(util_ep->rx_cq, &wc);
+	}
+}
+
+static inline
+void ofi_verbs_ud_ep_recv_multiple_progress(struct util_ep *util_ep)
+{
+	struct ofi_verbs_ud_ep *ep;
+	int num_ent, i;
+	struct ibv_wc *wcs;
+
+	ep = container_of(util_ep, struct ofi_verbs_ud_ep, util_ep);
+	wcs = ep->multiple_wcs;
+
+	num_ent = ibv_poll_cq(ep->ibv_send_cq,
+			      gl_data.num_poll_multiple_wcs,
+			      wcs);
+	for (i = 0; i < num_ent; i++) {
+		if (likely(wcs[i].status == IBV_WC_SUCCESS))
+			(void) ofi_verbs_ud_cq_comp(util_ep->rx_cq, &wcs[i]);
+		else
+			ofi_verbs_ud_cq_report_error(util_ep->rx_cq, &wcs[i]);
+	}
+}
```

Comment:
This function should be combined with the previous call.  Just declare a small number of wc on the stack (e.g. 8) and always just ask to process that many at a time.  There's no need to expose this level of micro-optimization to the admin.